src/Sarif/Visitors/PartitioningVisitor.cs

```diff
@@ -0,0 +1,454 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    /// <summary>
+    /// A visitor that partitions a specified SARIF log into a set of "partition logs."
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Two results are placed in the same partition log if and only if a specified "partition
+    /// function" returns the same value (the "partition value") for both results, except that
+    /// results whose partition value is null are discarded (not placed in any of the partition
+    /// logs). We refer to a set of results with the same partition value as a "partition."
+    /// </para>
+    /// <para>
+    /// Each run in the original log is partitioned independently. If an original run contains no
+    /// results from a given partition, the partition log for that partition does not contain a run
+    /// corresponding to that original run.
+    /// </para>
+    /// <para>
+    /// Each run in each partition log contains only those elements of run-level collections such
+    /// as Run.Artifacts that are relevant to the subset of results from that original run which
+    /// belong to that partition.
+    /// </para>
+    /// </remarks>
+    /// <typeparam name="T">
+    /// The type of the object returned from the partition function with which the visitor is
+    /// constructed. It must be a reference type so that null is a valid value. It must override
+    /// bool Equals(T other) so that two Ts can compare equal even if they are not reference equal.
+    /// </typeparam>
+    public class PartitioningVisitor<T> : SarifRewritingVisitor where T : class
+    {
+        // The name of a property in the property bag of each partition log which contains the
+        // partition value for the results in that log file.
+        private const string PartitionValuePropertyName = "partitionValue";
+
+        /// <summary>
+        /// A delegate for a function that returns a value specifying which partition a specified
+        /// result belongs to, or null if the result should be discarded (not placed in any
+        /// partition).
+        /// </summary>
+        /// <param name="result">
+        /// The result to be assigned to a partition.
+        /// </param>
+        /// <typeparam name="T">
+        /// The type of the object returned from the partition function. It must be a reference
+        /// type so that null is a valid value. It must override bool Equals(T other) so that
+        /// two Ts can compare equal even if they are not reference equal.
```

Comment:
Interesting point. According to [StackOverflow](https://stackoverflow.com/questions/15889987/when-to-specify-constraint-t-iequatablet-even-though-it-is-not-strictly-re), it doesn't actually add anything because all types are convertible to `System.Object`, which implements `IEquatable<T>`.

**BUT** The poster goes on to argue that it makes for good documentation, telling the user that the objects are going to be compared, so perhaps he should give some thought to how the comparison would work.

On the other other hand, my comment already does that.

TL;DR: I'll add it.

---
In reply to: [357427829](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357427829) [](ancestors = 357427829)